src/dezerialize.ts

```diff
@@ -235,60 +247,107 @@ const dezerializers = {
-          dezerialize(value),
+          dezerialize(value as any, opts),
         ])
       )
     )) as any,
-  record: ((shape: SzRecord) =>
-    z.record(dezerialize(shape.key), dezerialize(shape.value))) as any,
-  map: ((shape: SzMap<any, any>) =>
-    z.map(dezerialize(shape.key), dezerialize(shape.value))) as any,
+  record: ((shape: SzRecord, opts: DezerializerOptions) =>
+    z.record(
+      dezerialize(shape.key, opts),
+      dezerialize(shape.value, opts)
+    )) as any,
+  map: ((shape: SzMap<any, any>, opts: DezerializerOptions) =>
+    z.map(dezerialize(shape.key, opts), dezerialize(shape.value, opts))) as any,
 
   enum: ((shape: SzEnum) => z.enum(shape.values)) as any,
 
-  union: ((shape: SzUnion) =>
-    z.union(shape.options.map(dezerialize) as any)) as any,
-  discriminatedUnion: ((shape: SzDiscriminatedUnion) =>
+  union: ((shape: SzUnion, opts: DezerializerOptions) =>
+    z.union(shape.options.map((opt) => dezerialize(opt, opts)) as any)) as any,
+  discriminatedUnion: ((
+    shape: SzDiscriminatedUnion,
+    opts: DezerializerOptions
+  ) =>
     z.discriminatedUnion(
       shape.discriminator,
-      shape.options.map(dezerialize) as any
+      shape.options.map((opt) => dezerialize(opt, opts)) as any
+    )) as any,
+  intersection: ((shape: SzIntersection, opts: DezerializerOptions) =>
+    z.intersection(
+      dezerialize(shape.left, opts),
+      dezerialize(shape.right, opts)
     )) as any,
-  intersection: ((shape: SzIntersection) =>
-    z.intersection(dezerialize(shape.left), dezerialize(shape.right))) as any,
 
-  function: ((shape: SzFunction<any, any>) =>
+  function: ((shape: SzFunction<any, any>, opts: DezerializerOptions) =>
     z.function(
-      dezerialize(shape.args) as any,
-      dezerialize(shape.returns)
+      dezerialize(shape.args, opts) as any,
+      dezerialize(shape.returns, opts)
     )) as any,
-  promise: ((shape: SzPromise) => z.promise(dezerialize(shape.value))) as any,
+  promise: ((shape: SzPromise, opts: DezerializerOptions) =>
+    z.promise(dezerialize(shape.value, opts))) as any,
+  effect: ((shape: SzEffect, opts: DezerializerOptions) => {
+    let base = dezerialize(shape.inner, opts);
+    if (
+      !(
+        "superRefinements" in opts ||
+        "transforms" in opts ||
+        "preprocesses" in opts
+      )
+    ) {
+      return base;
+    }
+    for (const { name, type } of shape.effects) {
+      if (
+        type === "refinement" &&
+        opts.superRefinements &&
+        opts.superRefinements[name]
+      ) {
+        base = base.superRefine(opts.superRefinements[name]);
+      } else if (
+        type === "transform" &&
+        opts.transforms &&
+        opts.transforms[name]
+      ) {
+        base = base.transform(opts.transforms[name]);
+      } else if (
+        type === "preprocess" &&
+        opts.preprocesses &&
+        opts.preprocesses[name]
+      ) {
+        base = z.preprocess(opts.preprocesses[name], base);
+      }
+    }
```

Comment:
nit: I like this syntax for keeping things terser, wdyt?
```suggestion
    for (const { name, type } of shape.effects) {
      if (type === "refinement" && opts.superRefinements?.[name]) {
        base = base.superRefine(opts.superRefinements[name]);
      } else if (type === "transform" && opts.transforms?.[name]) {
        base = base.transform(opts.transforms[name]);
      } else if (type === "preprocess" && opts.preprocesses?.[name]) {
        base = z.preprocess(opts.preprocesses[name], base);
      }
    }
```

Reply:
Much better, thanks!